File: brain-move/games/brain_trainer.py
import random
import time
import math
from games.base_game import BaseGame


class BrainTrainer(BaseGame):
    def __init__(self, difficulty_engine, scoring_engine):
        super().__init__(difficulty_engine, scoring_engine)
        self.name = "brain_trainer"
        self.display_name = "Brain Trainer"
        self.phase = "instruction"
        self.round = 0
        self.total_rounds = 10
        self.sub_game = "math"
        self.correct_count = 0
        self.total_answered = 0
        self.current_question = {}
        self.streak = 0
        self.instruction_timer = 0
        self.next_timer = 0
        self.show_result = ""
        self.show_result_timer = 0
        self.SUB_GAMES = ["math", "recall", "count"]
        self.recall_items = []
        self.recall_answer = ""
        self.count_grid = []
        self.count_target = 0

    def start(self):
        super().start()
        self.phase = "instruction"
        self.round = 0
        self.correct_count = 0
        self.total_answered = 0
        self.streak = 0
        self.instruction_timer = time.time()
        self.feedback_message = "Brain Trainer: rispondi alle domande!"

    def update(self, pose_data=None, dt=0):
        if not self.running:
            return
        if self.phase == "instruction":
            if time.time() - self.instruction_timer > 2.5:
                self._next_round()
            return
        if self.phase == "result":
            self.show_result_timer -= dt
            if self.show_result_timer <= 0:
                self._next_round()

    def _next_round(self):
        if self.round >= self.total_rounds:
            self._finish_game()
            return
        self.round += 1
        self.show_result = ""
        self.phase = "playing"
        self.sub_game = random.choice(self.SUB_GAMES)
        if self.sub_game == "math":
            self._gen_math()
        elif self.sub_game == "recall":
            self._gen_recall()
        else:
            self._gen_count()

    def _gen_math(self):
        level = self.difficulty.get_level()
        max_num = 5 + level * 2
        a = random.randint(1, max_num)
        b = random.randint(1, max_num)
        op = random.choice(["+", "-"])
        if op == "-" and a < b:
            a, b = b, a
        answer = a + b if op == "+" else a - b
        wrongs = set()
        while len(wrongs) < 3:
            w = answer + random.randint(-3, 3)
            if w != answer and w >= 0:
                wrongs.add(w)
        options = [answer] + list(wrongs)
        random.shuffle(options)
        self.current_question = {
            "type": "math",
            "text": f"{a} {op} {b} = ?",
            "options": options,
            "answer": answer,
        }
        self.feedback_message = f"Round {self.round}/{self.total_rounds}: {a} {op} {b} = ?"

    def _gen_recall(self):
        level = self.difficulty.get_level()
        items = ["\U0001F34E", "\U0001F431", "\U0001F697", "\U0001F3E0",
                 "\U0001F31E", "\U0001F436", "\U0001F33C", "\u2B50",
                 "\U0001F30A", "\U0001F308", "\U0001F422", "\U0001F426"]
        count = min(3 + level, 6)
        self.recall_items = random.sample(items, count)
        self.recall_answer = random.choice(self.recall_items)
        self.current_question = {
            "type": "recall",
            "text": "Ricorda l'immagine cerchiata!",
            "items": self.recall_items,
            "highlight": self.recall_answer,
            "options": self.recall_items[:],
            "answer": self.recall_answer,
        }
        random.shuffle(self.current_question["options"])
        self.feedback_message = f"Round {self.round}: ricorda e scegli l'immagine giusta"

    def _gen_count(self):
        level = self.difficulty.get_level()
        count = 3 + level
        self.count_target = random.randint(3, count + 2)
        symbols = ["\U0001F535", "\U0001F7E2", "\U0001F534", "\U0001F7E1"]
        s = random.choice(symbols)
        self.count_grid = [s] * self.count_target
        options = [self.count_target]
        for _ in range(3):
            w = self.count_target + random.randint(-3, 3)
            if w > 0 and w not in options:
                options.append(w)
        options = options[:4]
        random.shuffle(options)
        self.current_question = {
            "type": "count",
            "text": "Quanti simboli ci sono?",
            "symbols": self.count_grid,
            "symbol_char": s,
            "options": options,
            "answer": self.count_target,
        }
        self.feedback_message = f"Round {self.round}: conta i simboli!"

    def answer(self, value):
        if self.phase != "playing":
            return
        q = self.current_question
        correct = (value == q["answer"])
        self.total_answered += 1
        if correct:
            self.correct_count += 1
            self.streak += 1
            self.show_result = "Corretto! +25 punti"
            self.scoring.add_score(self.name, 25)
        else:
            self.streak = 0
            self.show_result = f"Sbagliato! Era {q['answer']}"
        self.phase = "result"
        self.show_result_timer = 2.0

    def _finish_game(self):
        self.stop()
        accuracy = self.correct_count / max(self.total_answered, 1)
        self.result = {
            "completed": True,
            "accuracy": accuracy,
            "correct": self.correct_count,
            "total": self.total_answered,
            "streak": self.streak,
        }
        self.difficulty.adapt(accuracy)
        self.scoring.add_score(self.name, int(accuracy * 30))

    def get_progress(self):
        return self.round / max(self.total_rounds, 1)

    def get_question_options(self):
        return self.current_question.get("options", [])

File: brain-move-web/src/games/brain-trainer.ts
import { BaseGame } from './base-game'

export class BrainTrainer extends BaseGame {
  phase: "instruction" | "playing" | "result" = "instruction"
  round = 0
  totalRounds = 10
  subGame: "math" | "recall" | "count" = "math"
  correctCount = 0
  totalAnswered = 0
  currentQuestion: Record<string, unknown> = {}
  streak = 0
  instructionTimer = 0
  resultTimer = 0
  subGames = ["math", "recall", "count"]

  constructor(difficulty: any, scoring: any) {
    super(difficulty, scoring)
    this.name = "brain_trainer"
    this.displayName = "Brain Trainer"
  }

  start(): void {
    super.start()
    this.phase = "instruction"
    this.round = 0
    this.correctCount = 0
    this.totalAnswered = 0
    this.streak = 0
    this.instructionTimer = Date.now()
    this.feedbackMessage = "Brain Trainer: rispondi alle domande!"
  }

  update(_poseData?: any, dt = 0): void {
    if (!this.running) return
    if (this.phase === "instruction") {
      if ((Date.now() - this.instructionTimer) / 1000 > 2.5) this._nextRound()
      return
    }
    if (this.phase === "result") {
      this.resultTimer -= dt
      if (this.resultTimer <= 0) this._nextRound()
    }
  }

  private _nextRound(): void {
    if (this.round >= this.totalRounds) { this._finishGame(); return }
    this.round++
    this.phase = "playing"
    this.subGame = this.subGames[Math.floor(Math.random() * this.subGames.length)] as "math" | "recall" | "count"
    if (this.subGame === "math") this._genMath()
    else if (this.subGame === "recall") this._genRecall()
    else this._genCount()
  }

  private _genMath(): void {
    const level = this.difficulty.getLevel()
    const maxNum = 5 + level * 2
    let a = Math.floor(Math.random() * maxNum) + 1
    let b = Math.floor(Math.random() * maxNum) + 1
    const op = Math.random() < 0.5 ? "+" : "-"
    if (op === "-" && a < b) [a, b] = [b, a]
    const answer = op === "+" ? a + b : a - b
    const wrongs = new Set<number>()
    while (wrongs.size < 3) { const w = answer + Math.floor(Math.random() * 7) - 3; if (w !== answer && w >= 0) wrongs.add(w) }
    const options = [answer, ...wrongs].sort(() => Math.random() - 0.5)
    this.currentQuestion = { type: "math", text: `${a} ${op} ${b} = ?`, options, answer }
    this.feedbackMessage = `Round ${this.round}/${this.totalRounds}: ${a} ${op} ${b} = ?`
  }

  private _genRecall(): void {
    const level = this.difficulty.getLevel()
    const items = ["\u{1F34E}", "\u{1F431}", "\u{1F697}", "\u{1F3E0}", "\u{1F31E}", "\u{1F436}", "\u{1F33C}", "\u2B50", "\u{1F30A}", "\u{1F308}", "\u{1F422}", "\u{1F426}"]
    const count = Math.min(3 + level, 6)
    const recallItems = items.sort(() => Math.random() - 0.5).slice(0, count)
    const answer = recallItems[Math.floor(Math.random() * recallItems.length)]
    const options = [...recallItems].sort(() => Math.random() - 0.5)
    this.currentQuestion = { type: "recall", text: "Ricorda l'immagine cerchiata!", items: recallItems, highlight: answer, options, answer }
    this.feedbackMessage = `Round ${this.round}: ricorda e scegli l'immagine giusta`
  }

  private _genCount(): void {
    const level = this.difficulty.getLevel()
    const count = 3 + level
    const target = Math.floor(Math.random() * count) + 3
    const symbols = ["\u{1F535}", "\u{1F7E2}", "\u{1F534}", "\u{1F7E1}"]
    const s = symbols[Math.floor(Math.random() * symbols.length)]
    const grid = new Array(target).fill(s)
    const optionsSet = new Set<number>([target])
    while (optionsSet.size < 4) { const w = target + Math.floor(Math.random() * 7) - 3; if (w > 0) optionsSet.add(w) }
    const options = [...optionsSet].sort(() => Math.random() - 0.5).slice(0, 4)
    this.currentQuestion = { type: "count", text: "Quanti simboli ci sono?", symbols: grid, symbolChar: s, options, answer: target }
    this.feedbackMessage = `Round ${this.round}: conta i simboli!`
  }

  answer(value: unknown): void {
    if (this.phase !== "playing") return
    const q = this.currentQuestion
    const correct = value === q.answer
    this.totalAnswered++
    if (correct) {
      this.correctCount++; this.streak++
      this.feedbackMessage = "Corretto! +25 punti"
      this.scoring.addScore(this.name, 25)
    } else {
      this.streak = 0
      this.feedbackMessage = `Sbagliato! Era ${q.answer}`
    }
    this.phase = "result"
    this.resultTimer = 2.0
  }

  private _finishGame(): void {
    this.stop()
    const accuracy = this.correctCount / Math.max(this.totalAnswered, 1)
    this.result = { completed: true, accuracy, correct: this.correctCount, total: this.totalAnswered, streak: this.streak }
    this.difficulty.adapt(accuracy)
    this.scoring.addScore(this.name, Math.round(accuracy * 30))
  }

  getProgress(): number { return this.round / Math.max(this.totalRounds, 1) }
  getQuestionOptions(): unknown[] { return (this.currentQuestion.options as unknown[]) ?? [] }
  getCurrentQuestion(): Record<string, unknown> { return this.currentQuestion }
  getSubGame(): string { return this.subGame }
}
